File: shared/src/main/java/com/guardpulse/parentcontrol/shared/PinHasher.kt
package com.guardpulse.parentcontrol.shared

import java.security.MessageDigest
import java.security.SecureRandom
import java.util.Base64
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.PBEKeySpec

data class PinHash(
    val salt: String,
    val hash: String,
    val version: Int = PinHasher.CURRENT_VERSION,
    val algorithm: String = PinHasher.CURRENT_ALGORITHM,
    val iterations: Int = PinHasher.CURRENT_ITERATIONS
)

object PinHasher {
    const val LEGACY_VERSION = 1
    const val CURRENT_VERSION = 2
    const val CURRENT_ALGORITHM = "PBKDF2WithHmacSHA256"
    const val CURRENT_ITERATIONS = 210_000
    private const val KEY_LENGTH_BITS = 256
    private val random = SecureRandom()

    fun create(pin: String): PinHash {
        require(pin.matches(Regex("\\d{6}"))) { "PIN must be exactly six digits" }
        val saltBytes = ByteArray(16)
        random.nextBytes(saltBytes)
        val salt = Base64.getUrlEncoder().withoutPadding().encodeToString(saltBytes)
        return PinHash(salt, pbkdf2(pin, salt, CURRENT_ITERATIONS))
    }

    fun verify(
        pin: String,
        salt: String,
        expectedHash: String,
        version: Int = CURRENT_VERSION,
        algorithm: String? = null,
        iterations: Int? = null
    ): Boolean {
        if (!pin.matches(Regex("\\d{6}")) || salt.isBlank() || expectedHash.isBlank()) return false
        val actual = when (version) {
            LEGACY_VERSION -> legacyHash(pin, salt)
            CURRENT_VERSION -> {
                if (algorithm != null && algorithm != CURRENT_ALGORITHM) return false
                val rounds = iterations ?: CURRENT_ITERATIONS
                if (rounds !in CURRENT_ITERATIONS..1_000_000) return false
                pbkdf2(pin, salt, rounds)
            }
            else -> return false
        }
        return MessageDigest.isEqual(
            actual.toByteArray(Charsets.UTF_8),
            expectedHash.toByteArray(Charsets.UTF_8)
        )
    }

    internal fun createLegacyForTest(pin: String, salt: String): PinHash {
        return PinHash(salt, legacyHash(pin, salt), LEGACY_VERSION, "SHA-256", 1)
    }

    private fun pbkdf2(pin: String, salt: String, iterations: Int): String {
        val spec = PBEKeySpec(
            pin.toCharArray(),
            Base64.getUrlDecoder().decode(salt),
            iterations,
            KEY_LENGTH_BITS
        )
        return try {
            val bytes = SecretKeyFactory.getInstance(CURRENT_ALGORITHM).generateSecret(spec).encoded
            Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)
        } finally {
            spec.clearPassword()
        }
    }

    private fun legacyHash(pin: String, salt: String): String {
        val digest = MessageDigest.getInstance("SHA-256")
        val bytes = digest.digest("$salt:$pin".toByteArray(Charsets.UTF_8))
        return Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)
    }
}

File: web/js/pinHasher.js
/**
 * Port of shared/PinHasher.kt — byte-identical hashes via WebCrypto.
 * v2: PBKDF2-HMAC-SHA256, 210,000 iterations, 16-byte salt, 32-byte key,
 *     base64url unpadded.
 * v1 (legacy, verify-only): SHA-256 of UTF-8 "salt:pin", base64url unpadded.
 */

export const LEGACY_VERSION = 1;
export const CURRENT_VERSION = 2;
export const CURRENT_ALGORITHM = 'PBKDF2WithHmacSHA256';
export const CURRENT_ITERATIONS = 210_000;
const KEY_LENGTH_BITS = 256;

export function assertPinFormat(pin) {
  if (!/^\d{6}$/.test(pin)) throw new Error('PIN must be exactly six digits');
}

export async function create(pin) {
  assertPinFormat(pin);
  const saltBytes = crypto.getRandomValues(new Uint8Array(16));
  const salt = bytesToBase64Url(saltBytes);
  return {
    salt,
    hash: await pbkdf2(pin, salt, CURRENT_ITERATIONS),
    version: CURRENT_VERSION,
    algorithm: CURRENT_ALGORITHM,
    iterations: CURRENT_ITERATIONS,
  };
}

export async function verify(pin, salt, expectedHash, version = CURRENT_VERSION, algorithm = null, iterations = null) {
  if (!/^\d{6}$/.test(pin) || !salt || !expectedHash) return false;
  let actual;
  if (version === LEGACY_VERSION) {
    actual = await legacyHash(pin, salt);
  } else if (version === CURRENT_VERSION) {
    if (algorithm != null && algorithm !== CURRENT_ALGORITHM) return false;
    const rounds = iterations ?? CURRENT_ITERATIONS;
    if (rounds < CURRENT_ITERATIONS || rounds > 1_000_000) return false;
    actual = await pbkdf2(pin, salt, rounds);
  } else {
    return false;
  }
  // Kotlin compares the UTF-8 bytes of the base64 strings (MessageDigest.isEqual).
  return constantTimeEqualUtf8(actual, expectedHash);
}

async function pbkdf2(pin, salt, iterations) {
  const keyMaterial = await crypto.subtle.importKey(
    'raw', new TextEncoder().encode(pin), 'PBKDF2', false, ['deriveBits'],
  );
  const bits = await crypto.subtle.deriveBits(
    { name: 'PBKDF2', hash: 'SHA-256', salt: base64UrlToBytes(salt), iterations },
    keyMaterial,
    KEY_LENGTH_BITS,
  );
  return bytesToBase64Url(new Uint8Array(bits));
}

async function legacyHash(pin, salt) {
  const digest = await crypto.subtle.digest('SHA-256', new TextEncoder().encode(`${salt}:${pin}`));
  return bytesToBase64Url(new Uint8Array(digest));
}

function constantTimeEqualUtf8(a, b) {
  const aBytes = new TextEncoder().encode(a);
  const bBytes = new TextEncoder().encode(b);
  if (aBytes.length !== bBytes.length) return false;
  let diff = 0;
  for (let i = 0; i < aBytes.length; i++) diff |= aBytes[i] ^ bBytes[i];
  return diff === 0;
}

function bytesToBase64Url(bytes) {
  let binary = '';
  bytes.forEach((b) => { binary += String.fromCharCode(b); });
  return btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/g, '');
}

function base64UrlToBytes(url) {
  let b64 = url.replace(/-/g, '+').replace(/_/g, '/');
  while (b64.length % 4 !== 0) b64 += '=';
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) bytes[i] = binary.charCodeAt(i);
  return bytes;
}
